coderd/devtunnel/tunnel.go

```diff
@@ -1,93 +1,270 @@
+			_ = wgListen.Close()
+			dev.Close()
+			close(ch)
+
+		case <-dev.Wait():
+			close(ch)
+		}
 	}()
+
+	return &Tunnel{
+		URL:       fmt.Sprintf("https://%s.%s", cfg.ID, EndpointHTTPS),
+		Listener:  wgListen,
+		ErrorChan: ch,
+	}, nil
+}
+
+// New creates a tunnel with a public URL and returns a listener for incoming
+// connections on that URL. Connections are made over the wireguard protocol.
+// Tunnel configuration is cached in the user's config directory. Successive
+// calls to New will always use the same URL. If multiple public URLs in
+// parallel are required, use NewWithConfig.
+func New(ctx context.Context, logger slog.Logger) (*Tunnel, error) {
+	cfg, err := readOrGenerateConfig()
+	if err != nil {
+		return nil, xerrors.Errorf("read or generate config: %w", err)
+	}
+
+	return NewWithConfig(ctx, logger, cfg)
+}
+
+func startUpdateRoutine(ctx context.Context, logger slog.Logger, cfg Config) error {
+	// Ensure we send the first config before spawning in the background.
+	_, err := sendConfigToServer(ctx, cfg)
+	if err != nil {
+		return xerrors.Errorf("send config to server: %w", err)
+	}
+
 	go func() {
-		<-ctx.Done()
-		svc.Close()
+		ticker := time.NewTicker(30 * time.Second)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				break
+			case <-ticker.C:
+			}
+
+			_, err := sendConfigToServer(ctx, cfg)
+			if err != nil {
+				logger.Debug(ctx, "send tunnel config to server", slog.Error(err))
+			}
+		}
 	}()
+	return nil
+}
+
+func sendConfigToServer(_ context.Context, cfg Config) (created bool, err error) {
+	raw, err := json.Marshal(configExt(cfg))
+	if err != nil {
+		return false, xerrors.Errorf("marshal config: %w", err)
+	}
+
+	res, err := http.Post("https://"+EndpointHTTPS+"/tun", "application/json", bytes.NewReader(raw))
+	if err != nil {
+		return false, xerrors.Errorf("send request: %w", err)
+	}
+	_, _ = io.Copy(io.Discard, res.Body)
+	_ = res.Body.Close()
+
+	return res.StatusCode == http.StatusCreated, nil
+}
+
+func cfgPath() (string, error) {
+	cfgDir, err := os.UserConfigDir()
+	if err != nil {
+		return "", xerrors.Errorf("get user config dir: %w", err)
+	}
+
+	return filepath.Join(cfgDir, "coderv2", "devtunnel"), nil
+}
+
+func readOrGenerateConfig() (Config, error) {
+	cfgFi, err := cfgPath()
+	if err != nil {
+		return Config{}, xerrors.Errorf("get config path: %w", err)
+	}
+
+	fi, err := os.ReadFile(cfgFi)
+	if err != nil {
+		if os.IsNotExist(err) {
+			cfg, err := GenConfig()
+			if err != nil {
+				return Config{}, xerrors.Errorf("generate config: %w", err)
+			}
+
+			err = writeConfig(cfg)
+			if err != nil {
+				return Config{}, xerrors.Errorf("write config: %w", err)
+			}
+
+			return cfg, nil
+		}
+
+		return Config{}, xerrors.Errorf("read config: %w", err)
+	}
+
+	cfg := Config{}
+	err = json.Unmarshal(fi, &cfg)
+	if err != nil {
+		return Config{}, xerrors.Errorf("unmarshal config: %w", err)
+	}
+
+	return cfg, nil
+}
+
+func GenConfig() (Config, error) {
```

Comment:
I might expand this to `GenerateConfig` to be verbose like `GeneratePrivateKey` used inside the function, but this is a preference thing so chefs choice!